Mathlib/Tactic/finCases.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2022 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison, Hanting Zhang
+-/
+
+import Mathlib.Lean.Expr.Basic
+import Mathlib.Data.Fintype.Basic
+
+open Lean.Meta
+
+namespace Lean.Elab.Tactic
+
+def getMemType (e : Expr) : TacticM (Option Expr) := do
+  -- let type ← inferType e -- should be `Prop`
+  match e.getAppFnArgs with
+  | (``Membership.mem, #[_, type, _, _, _]) =>
+    match type.getAppFnArgs with
+    | (``List, #[α])     => return α
+    | (``Multiset, #[α]) => return α
+    | (``Finset, #[α])   => return α
+    | _ => throwError m!"Hypothesis must be of type `x ∈ (A : List/Multiset/Finset α)`"
+  | _ => return none
+
+partial def finCasesAt (hyp : FVarId) : TacticM Unit := do
+  withMainContext do
+    let lDecl ←
+      match (← getLCtx).find? hyp with
+      | none => throwError m!"hypothesis not found"
+      | some lDecl => pure lDecl
+    match ← getMemType lDecl.type with
+    | some _ =>
+      -- Deal with `x ∈ A` hypotheses:
+      -- everything inside this block is a hack lmao
+      liftMetaTacticAux fun mvarId => do
```

Comment:
This feels like a gigantic hack. Is there a way recursively apply `cases` and capture the subgoals in a list nicely?